internal/txmanager/txmanager_test.go

```diff
@@ -0,0 +1,250 @@
+			s.Require().NoError(err)
+
+			_, err = tx.ExecContext(ctx, `UPDATE dummy SET value = 'tx1' WHERE id = 'id1'`)
+			s.Require().NoError(err)
+
+			ch1 <- struct{}{}
+			<-ch2
+
+			return nil
+		})
+		s.Require().NoError(err)
+	}()
+
+	go func() {
+		err := s.txMgr.WithTransaction(s.ctx, func(ctx context.Context) error {
+			tx, err := txmanager.GetTx(ctx)
+			s.Require().NoError(err)
+
+			<-ch1
+			_, err = tx.ExecContext(ctx, `UPDATE dummy SET value = 'tx2' WHERE id = 'id2'`)
+			s.Require().NoError(err)
+
+			ch2 <- struct{}{}
+			return nil
+		})
+		s.Require().NoError(err)
+	}()
+
+	time.Sleep(2 * time.Second)
+
+	var v1, v2 string
+	s.Require().NoError(s.db.Get(&v1, `SELECT value FROM dummy WHERE id = 'id1'`))
+	s.Require().NoError(s.db.Get(&v2, `SELECT value FROM dummy WHERE id = 'id2'`))
+	s.Equal("tx1", v1)
+	s.Equal("tx2", v2)
+}
+
+func (s *TxManagerTestSuite) TestParallelTransactions_Conflicting() {
+	_, err := s.db.ExecContext(s.ctx, `
+        CREATE TABLE IF NOT EXISTS dummy (
+            id TEXT PRIMARY KEY,
+            value TEXT NOT NULL
+        );
+    `)
+	s.Require().NoError(err)
+
+	_, err = s.db.ExecContext(s.ctx, `DELETE FROM dummy;`)
+	s.Require().NoError(err)
+
+	_, err = s.db.ExecContext(s.ctx, `INSERT INTO dummy (id, value) VALUES ('id1', 'init');`)
+	s.Require().NoError(err)
+
+	chReady := make(chan struct{})
+	chContinue := make(chan struct{})
+	errCh := make(chan error, 2)
+
+	// Tx1
+	go func() {
+		err := s.txMgr.WithTransaction(s.ctx, func(ctx context.Context) error {
+			tx, err := txmanager.GetTx(ctx)
+			s.Require().NoError(err)
+
+			s.T().Log("Tx1: reading row")
+			var val string
+			err = tx.GetContext(ctx, &val, `SELECT value FROM dummy WHERE id = 'id1'`)
+			s.Require().NoError(err)
+
+			s.T().Log("Tx1: signaling Tx2 to start")
+			chReady <- struct{}{}
+
+			s.T().Log("Tx1: waiting for Tx2 to read and write")
+			<-chContinue
+
+			s.T().Log("Tx1: updating row")
+			_, err = tx.ExecContext(ctx, `UPDATE dummy SET value = 'tx1' WHERE id = 'id1'`)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		})
+		s.T().Logf("Tx1 done: %v", err)
+		errCh <- err
+	}()
+
+	// Tx2
+	go func() {
+		err := s.txMgr.WithTransaction(s.ctx, func(ctx context.Context) error {
+			tx, err := txmanager.GetTx(ctx)
+			s.Require().NoError(err)
+
+			s.T().Log("Tx2: waiting for Tx1 to read")
+			<-chReady
+
+			s.T().Log("Tx2: reading row")
+			var val string
+			err = tx.GetContext(ctx, &val, `SELECT value FROM dummy WHERE id = 'id1'`)
+			s.Require().NoError(err)
+
+			s.T().Log("Tx2: updating row")
+			_, err = tx.ExecContext(ctx, `UPDATE dummy SET value = 'tx1' WHERE id = 'id1'`)
+			if err != nil {
+				return err
+			}
+
+			s.T().Log("Tx2: signaling Tx1 to continue")
+			chContinue <- struct{}{}
+
+			return nil
+		})
+		s.T().Logf("Tx2 done: %v", err)
+		errCh <- err
+	}()
+
+	err1 := <-errCh
+	err2 := <-errCh
+
+	s.T().Logf("Tx1 error: %v", err1)
+	s.T().Logf("Tx2 error: %v", err2)
```

Comment:
You can not know for sure whether `err1` originates in Tx1 and `err2` in Tx2. They could have been written to the channel in arbitrary order. What you could do is:
1. Synchronise writes to `errCh`, just as you do with `chReady` and `chContinue`, to enforce specific order
2. Along with errors write in which transactions they have originated in. So make `errCh` a channel of structures with `TxIdx` and `error`
3. Create `errs := make([]error, 2)` and fill `errs[0]` in Tx1 and `errs[1]` in Tx2. It is thread-safe to concurrently change slice entries, provided that the slice itself wont reallocate (so no appends).

Reply:
https://github.com/Albertthea/student-service/pull/8#pullrequestreview-3026286279